core/templates/dev/head/pages/exploration_player/PretestEngineService.js

```diff
@@ -0,0 +1,276 @@
+    var _explorationId = ContextService.getExplorationId();
+
+    var version = GLOBALS.explorationVersion;
+
+    var answerIsBeingProcessed = false;
+
+    var pretestQuestions = [];
+
+    var currentIndex = null;
+    var nextIndex = null;
+
+    var randomFromArray = function(arr) {
+      return arr[Math.floor(Math.random() * arr.length)];
+    };
+
+    // Evaluate feedback.
+    var makeFeedback = function(feedbackHtml, envs) {
+      return ExpressionInterpolationService.processHtml(feedbackHtml, envs);
+    };
+
+    // Evaluate question string.
+    var makeQuestion = function(newState, envs) {
+      return ExpressionInterpolationService.processHtml(
+        newState.content.getHtml(), envs);
+    };
+
+    // This should only be called when 'exploration' is non-null.
+    var _loadInitialQuestion = function(successCallback) {
+      var initialState = pretestQuestions[0].getStateData();
+
+      var questionHtml = makeQuestion(initialState, []);
+      if (questionHtml === null) {
+        AlertsService.addWarning('Expression parsing error.');
+        return;
+      }
+
+      currentIndex = 0;
+      nextIndex = 0;
+      successCallback(null, questionHtml);
+    };
+
+    var _getCurrentStateData = function() {
+      return pretestQuestions[currentIndex].getStateData();
+    };
+
+    var _getNextStateData = function() {
+      return pretestQuestions[nextIndex].getStateData();
+    };
+
+    return {
+      /**
+       * Initializes the pretests for an exploration, passing the data for the
+       * first question to successCallback.
+       *
+       *
+       * @param {function} successCallback - The function to execute after the
+       *   pretest question data is successfully loaded. This function will
+       *   be passed two arguments:
+       *   - initHtml {string}, an HTML string representing the content of the
+       *       first state.
+       */
+      init: function(pretestQuestionDicts, successCallback) {
+        answerIsBeingProcessed = false;
+        for (var i = 0; i < pretestQuestionDicts.length; i++) {
+          pretestQuestions.push(
+            QuestionObjectFactory.createFromBackendDict(pretestQuestionDicts[i])
+          );
+        }
+        _loadInitialQuestion(successCallback);
+      },
+      recordNewCardAdded: function() {
+        currentIndex = nextIndex;
+      },
+      getPretestQuestionCount: function() {
+        return pretestQuestions.length;
+      },
+      getExplorationId: function() {
+        return _explorationId;
+      },
+      getExplorationVersion: function() {
+        return version;
+      },
+      getStateContentHtml: function() {
+        return _getCurrentStateData().content.getHtml();
+      },
+      getStateContentAudioTranslations: function() {
+        return null;
+      },
+      isContentAudioTranslationAvailable: function() {
+        return false;
+      },
+      getCurrentInteractionHtml: function(labelForFocusTarget) {
```

Comment:
Some of these methods should be defined on the card object, e.g. stateCard.getInteractionHtml(), stateCard.getAudioTranslations(), stateCard.getNextInteractionHtml(), stateCard.getInteraction(), etc.

Reply:
I thought the engine services shouldn't know the card object stored in the transcript?
Also, even if I do this, a racecondition can occur as the card object is added only in _addNewCard in conversation skin but there are various functions that call current and next interaction properties before that.
When navigating through history, I am using state card properties.

Reply:
Created an [issue](https://github.com/oppia/oppia/issues/5517) as discussed to handle the order of adding a new card.